tobac/utils/general.py

```diff
@@ -553,3 +554,129 @@ def combine_tobac_feats(list_of_feats, preserve_old_feat_nums=None):
+        The maximum horizontal distance (in meters) to transform features to.
+
+    Returns
+    -------
+    transformed_features: pd.DataFrame
+        A new feature dataframe, with the coordinates transformed to
+        the new grid, suitable for use in segmentation
+
+    """
+    from .. import analysis as tb_analysis
+
+    lat_coord, lon_coord = internal_utils.detect_latlon_coord_name(
+        new_dataset, latitude_name=latitude_name, longitude_name=longitude_name
+    )
+
+    if lat_coord not in features or lon_coord not in features:
+        raise ValueError("Cannot find latitude and/or longitude coordinate")
+
+    lat_vals_new = new_dataset[lat_coord].values
+    lon_vals_new = new_dataset[lon_coord].values
+
+    if len(lat_vals_new.shape) != len(lon_vals_new.shape):
+        raise ValueError(
+            "Cannot work with lat/lon coordinates of unequal dimensionality"
+        )
+
+    # the lat/lons must be a 2D grid, so if they aren't, make them one.
+    if len(lat_vals_new.shape) == 1:
+        lon_vals_new, lat_vals_new = np.meshgrid(lon_vals_new, lat_vals_new)
+
+    # we have to convert to radians because scikit-learn's haversine
+    # requires that the input be in radians.
+    flat_lats = np.deg2rad(lat_vals_new.flatten())
+    flat_lons = np.deg2rad(lon_vals_new.flatten())
+
+    # we have to drop NaN values.
+    either_nan = np.logical_or(np.isnan(flat_lats), np.isnan(flat_lons))
+    # we need to remember where these values are in the array so that we can
+    # appropriately unravel them.
+    loc_arr = np.arange(0, len(flat_lats), 1)
+    loc_arr_trimmed = loc_arr[~either_nan]
+    flat_lats_nona = flat_lats[~either_nan]
+    flat_lons_nona = flat_lons[~either_nan]
+    ll_tree = sklearn.neighbors.BallTree(
+        np.array([flat_lats_nona, flat_lons_nona]).T, metric="haversine"
+    )
+    new_h1 = list()
+    new_h2 = list()
+    # there is almost certainly room for speedup in here.
+    for index in features["index"]:
+        dist, closest_pt = ll_tree.query(
+            [
+                [
+                    np.deg2rad(features["latitude"][index]),
+                    np.deg2rad(features["longitude"][index]),
+                ]
+            ]
+        )
+        unraveled_h1, unraveled_h2 = np.unravel_index(
+            loc_arr_trimmed[closest_pt[0][0]], np.shape(lat_vals_new)
+        )
+
+        new_h1.append(unraveled_h1)
+        new_h2.append(unraveled_h2)
+    ret_features = copy.deepcopy(features)
+    ret_features["hdim_1"] = ("index", new_h1)
+    ret_features["hdim_2"] = ("index", new_h2)
+
+    # find where distances are too large and drop them.
+    new_lat = lat_vals_new[new_h1, new_h2]
+    new_lon = lon_vals_new[new_h1, new_h2]
+    dist_apart = (
+        tb_analysis.haversine(
+            new_lat, new_lon, features[lat_coord], features[lon_coord]
+        )
+        * 1000.0
+    )
+    ret_features = ret_features.where(dist_apart < max_space_away, drop=True)
```

Comment:
This drops features silently, would it be good to add a warning or logging for when features are dropped due to exceeding the distance threshold so as to make this behaviour clear?

Reply:
I am disinclined from adding warning messages as they can get quite expensive for large datasets. It would be better to have a flag in the function to keep (but disable in some way) the output features, but we would need to add the ability in segmentation to detect these bad features and not attempt to segment on them. I'm hesitant to dig into that rabbit hole for now.

Reply:
Would this function be called more than once for a single dataset? Would it be possible instead to do the accounting at the end and raise a single warning message informing the user that some features have been dropped?

Reply:
This is a good point, and not an idea that I had considered. I've added a warning that runs at the end. I have made it optional (but on by default).